### PR TITLE
Playar byound accunt age storid in db

### PR DESCRIPTION
**Pull request by Jordie0608**
Based upon #26350. `set_client_age_from_db()` and `sync_client_with_db()` have been merged. New clients are now added to the user table in a separate query than the one used to update their details upon connection; their player and account age is then calculated with DATEDIFF.
The code and regex used in `findJoinDate()` was changed a bit.

### CODE REVIEW

**Comment by Cyberboss** [comment]
>byound

**Comment by ExcessiveUseOfCobblestone** [comment]
This title makes me mad.

**Comment by Jordie0608** [comment]
Just pretend it's french.

**Comment by optimumtact** [comment]
uses two queries every connection instead of one 😞

**Comment by Cyberboss** [comment]
```
<Cyberboss> pls rename jordie's pr
<Cyberboss> is triggering
<%oranges> k I will
<+TheGhostOfWhibyl1> [tgstation] optimumtact closed pull request #26425: Playar byound accunt age storid in db (master...nodatabaseforoldmen) https://github.com/tgstation/tgstation/pull/26425
<+TheGhostOfWhibyl1> [tgstation] optimumtact reopened pull request #26425: Playar byound accunt age storid in db (master...nodatabaseforoldmen) https://github.com/tgstation/tgstation/pull/26425
<%oranges> Fixed
<Cyberboss> AHHH
```

juice when?

**Comment by RandomMarine** [comment]
DAY OF THE BLENDER

**Comment by Poojawa** [comment]
Sorry @Cyberboss the juicero is useless.

**Comment by Jordie0608** [comment]
@optimumtact Well it's down from the current 5 every connection.

**Comment by TalkingCactus** [comment]
Is there anywhere to access the `related_accounts_[cid/ip]` information somewhere else? It's useful for tracking ban evaders.

**Comment by Jordie0608** [comment]
Not in-game feasibly; we have two webtools that are better at doing this: https://github.com/tgstation/tgstation13.org and https://github.com/nfreader/newSS13tools

File: code/modules/client/client_procs.dm [hunk 1]
```diff
@@ -371,70 +372,72 @@ GLOBAL_LIST(external_rsc_urls)
...
+	var/datum/DBQuery/query_client_in_db = SSdbcore.NewQuery("SELECT 1 FROM [format_table_name("player")] WHERE ckey = '[sql_ckey]'")
+	if(!query_client_in_db.Execute())
 		return
-
-	//Logging player access
-
+	if(!query_client_in_db.NextRow())
+		new_player = 1
+		account_join_date = findJoinDate()
+		var/datum/DBQuery/query_add_player = SSdbcore.NewQuery("INSERT INTO [format_table_name("player")] VALUES ('[sql_ckey]', Now(), Now(), INET_ATON('[sql_ip]'), '[sql_computerid]', '[sql_admin_rank]', account_join_date ? \'[account_join_date]\' : NULL)")
```
Comment (optimumtact) [inline code/modules/client/client_procs.dm]: is it theoretically possible for someone, especially an admin, to edit the account_join date before it gets inserted? if so they could sql inject attack.

**Comment by Niknakflak** [comment]
I'll miss those variables

File: code/modules/client/client_procs.dm [hunk 2]
```diff
@@ -371,70 +372,73 @@ GLOBAL_LIST(external_rsc_urls)
...
+		var/datum/DBQuery/query_log_player = SSdbcore.NewQuery("UPDATE [format_table_name("player")] SET lastseen = Now(), ip = INET_ATON('[sql_ip]'), computerid = '[sql_computerid]', lastadminrank = '[sql_admin_rank]', accountjoindate = [account_join_date ? "'[account_join_date]'" : "NULL"] WHERE ckey = '[sql_ckey]'")
+		if(!query_log_player.Execute())
+			return
+	if(!account_join_date)
+		account_join_date = "Error"
 	var/datum/DBQuery/query_log_connection = SSdbcore.NewQuery("INSERT INTO `[format_table_name("connection_log")]` (`id`,`datetime`,`server_ip`,`server_port`,`ckey`,`ip`,`computerid`) VALUES(null,Now(),INET_ATON('[world.internet_address]'),'[world.port]','[sql_ckey]',INET_ATON('[sql_ip]'),'[sql_computerid]')")
 	query_log_connection.Execute()
 
+/client/proc/findJoinDate()
+	var/http[] = world.Export("http://byond.com/members/[ckey]?format=text")
```
Comment (lzimann) [inline code/modules/client/client_procs.dm]: any special reason on why this is `http[]` and not just `list/http`?(i know this is just moving from another file, but curious if there's anything on why)

Comment (Jordie0608) [inline code/modules/client/client_procs.dm]: ¯\_(ツ)_/¯
Should it be?

Comment (duncathan) [inline code/modules/client/client_procs.dm]: casting it to `/list` would be nice but if you don't have anything else to change there's no need. the main list operations (read and write) work without casting so

Comment (lzimann) [inline code/modules/client/client_procs.dm]: Well byond has no difference between [] and /list, it's just that we don't use this type of declaration anymore.

**Comment by MrStonedOne** [comment]
>related_accounts_ip and _cid removed since they didn't actually do anything.

Nope.

Unrelated change, get it out of this pr and into another.

File: code/modules/client/client_defines.dm [hunk 3]
```diff
@@ -39,9 +39,9 @@
 		////////////////////////////////////
 		//things that require the database//
 		////////////////////////////////////
-	var/player_age = "Requires database"	//So admins know why it isn't working - Used to determine how old the account is - in days.
-	var/related_accounts_ip = "Requires database"	//So admins know why it isn't working - Used to determine what other accounts previously logged in from this ip
-	var/related_accounts_cid = "Requires database"	//So admins know why it isn't working - Used to determine what other accounts previously logged in from this computer id
```
Comment (MrStonedOne) [inline code/modules/client/client_defines.dm]: Keep these

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]

File: code/modules/client/client_procs.dm (hunk 2 above)
Comment (duncathan) [inline code/modules/client/client_procs.dm]: I guess [] does cast it, weird

Comment (Cyberboss) [inline code/modules/client/client_procs.dm]: Can we just use `list` to keep the standard, please?

**Comment by Jordie0608** [comment]
Is that just because you think it should be in a different PR or because you'd prefer them to be kept? I removed them when obsoleting `sync_client_with_db()` as I would think most utilize our webtools.
If you'd rather them not removed I won't bother making a new PR for it.

**Comment by MrStonedOne** [comment]
Its that i don't think they should be removed, and we already have one downstream who wants them kept

In fact adding them to the player panel would be a worthwhile thing to do.

**Comment by lzimann** [comment]
The idea of displaying in the player panel is nice. I can do it if no one else does before.

**Comment by AnturK** [comment]
Conflicts.

**Comment by MrStonedOne** [comment]
The sql changes have been made (mostly, i kept the id column and gave it a unique index to keep it compatible with old code)

This can be merged at any time, at that point i'll remove the id column

**Comment by optimumtact** [comment]
good stuff

**Comment by MrStonedOne** [comment]
client New is all fucked:
* cid randomzier checks are failing and not redirecting
* byond account age isn't even working at all
*  some people just can't connect
* oh, and the ready up button broke for random people

I couldn't github revert it, so i reverted the update

**Pull request merged**